### ISSUE TITLE
Delete binary files created by delve after closing the debug session

### ISSUE THREAD

**Issue by ramya-rao-a** [Microsoft/vscode-go#1345]
_From @firelizzard18 on June 20, 2017 7:56_

- VSCode Version: Code 1.13.1 (379d2efb5539b09112c793d3d9a413017d736f89, 2017-06-14T18:13:05.928Z)
- OS Version: Darwin x64 16.6.0
- Extensions: 

|Extension|Author|Version|
|---|---|---|
|aurelia|AureliaEffect|0.3.4|
|vscode-npm-script|eg2|0.2.0|
|task-master|ianhoney|0.1.39|
|quicktask|lkytal|2.8.0|
|Go|lukehoban|0.6.62|
|debugger-for-chrome|msjsdiag|3.1.4|;

---

Steps to Reproduce:

1. Open a Golang project
2. Debug something
3. Stop debugging (via `Stop`, aka the red square or Shift+F5)

---

`delve` is the debugger for golang, or at least it's the one that the Go extension uses. When you debug something with delve, it creates a large binary file in the current directory. If you debug a main function (`dlv debug`), you get 'debug'. If you debug a test function (`dlv test`), you get 'debug.test'.

In normal `delve` usage, when you're done, you `quit` delve. Delve then deletes this file. Apparently VSCode gracelessly terminates (SIGKILL?) delve, which means the file sticks around.

_Copied from original issue: Microsoft/vscode#29089_

**Comment by ramya-rao-a** [comment]
_From @firelizzard18 on June 20, 2017 7:59_

This may be a regression from Microsoft/vscode-go#765 which fixed Microsoft/vscode-go#438

**Comment by ramya-rao-a** [comment]
_From @weinand on June 20, 2017 10:3_

Yes, VS Code terminates debug adapters "gracelessly" (and we are planning to make this more graceful in the future).
But for now the VS Code Go debugger could 'quit' delve nicely when receiving a `DisconnectRequest` from VS Code. This is always the last request sent from VS Code to the debug adapter before killing the debug adapter. And VS Code kills the VS Code Go debugger only after it has received the response from the `DisconnectRequest`. So the debug adapter can delay the killing enough so that delve could do its cleanup.

**Comment by NexWeb** [comment]
There is a better cleaner solution: https://github.com/NexWeb/vscode-go/blob/1cdf2ce2a133623922acd8199411c9f67ca44931/src/debugAdapter/goDebug.ts#L382

**Comment by chuchiring** [comment]
can I config the "debug" file path? for example I want dlv create the big debug file in my TEMP directory

**Comment by ecbrodie** [comment]
I just installed the latest version of VSCode, vscode-go and golang onto my machine (Mac OS 10.13.4, also latest). I'm still running into this issue.

I guess I'll `.gitignore` the `debug` file, but is anyone else still running into this issue? Ideally this is cleaned up automatically. I'm not excited to have to run a looping cleanup script.

**Comment by NexWeb** [comment]
Better to replace the mess of ineffective code in your terminateProcess.sh with what actually works, pkill debug. WFM, for more than a year without issue. The only hassle is having to re-modify terminateProcess.sh after updates of the extension. 

![pkill-debug](https://user-images.githubusercontent.com/3054475/39176123-00346998-477a-11e8-90f1-2ac080f797f8.jpg)

**Comment by mxschmitt** [comment]
@ramya-rao-a Are there any updates related to this issue?

**Comment by chuchiring** [comment]
@ramya-rao-a I tried but failed... or something wrong?

in my package.json, i modified the path:

							"output": {
								"type": "string",
								"description": "Output path for the binary of delve",
								"default": "r://" 
							},

but the DEBUG file still generate in path same as the .go file

**Comment by jadchaar** [comment]
Any updates on this? Been running into this and was wondering if it is in the pipeline for July or August.

**Comment by chuchiring** [comment]
@ramya-rao-a ok thank you, it worked after I created a launch.json and add my custom file path to the property "output"!, finally, almost half year.

**Comment by ramya-rao-a** [comment]
@NexWeb After looking into dlv apis, I found a command to detach gracefully that cleans up the debug binary that is created as well as kills all the processes spawned in the process. This means we no longer have to specifically kill any processes from our end. I know your fix as worked flawlessly for you for more than a year, but I would appreciate it if you could try the current fix. If you do decide to try it out, please follow the instructions in 
https://github.com/Microsoft/vscode-go/issues/1345#issuecomment-406125202

**Comment by ramya-rao-a** [comment]
The fix for this issue is now out in the latest update (0.6.85) to the Go extension.

**Comment by mvrhov** [comment]
I hate to be a nag, and I know that this was a regression fix, but some of us who came in late this fix is now a regression. IMO, the behavior as was before ws correct. every IDE that produces the executable leaves it in it's place.
The other thing is that now one to run the executable outside of the ide now must do a manual build. Also if the output dir & executable name are the same for the build script and the vscode debug the next run will remove the file.

**Comment by firelizzard18** [comment]
@mvrhov Do you normally want your executables to be named ‘debug’?

Every other IDE is designed for other languages. Go does things differently. For example, the convention of every other language is to put artifacts in their own folder. Go only generates one artifact (if you ignore the pkg folder) and it puts it in the current directory. Another example, the accepted convention for production builds is to use `go install` to install the artifact to the go bin folder, which you can add to your path. Not to mention that building go is stupid simple and stupid fast.

In summary, Go’s philosophy and tooling is radically different from essentially every other language, so it’s not a terribly valid comparison.

Plus VSCode isn’t an IDE in my book. A development environment, yes. But modular rather than ‘integrated’ (aka monolithic and bloated).

**Comment by mvrhov** [comment]
MY executable IS NOT NAMED debug. I have the `output` set under the configurations

**Comment by ramya-rao-a** [comment]
@mvrhov 

> every IDE that produces the executable leaves it in it's place.

Just to be on the same page, are you referring to the `debug` binary created by delve during the debugging process or the executable created when you run `go install`?

> The other thing is that now one to run the executable outside of the ide now must do a manual build

The `debug` binary created by delve was never meant to be run on its own. It is specifically for delve to use for debugging purposes. If you were to use delve from the terminal (`dlv debug`), you will see the same `debug` file being created for the duration of the debugging session. When you exit the debugging session, the file is deleted by delve.

You can also run the `Go: Install Current Package` command (which does nothing but `go install` on your pacakge)  to get a binary under GOPATH/bin

> the output dir & executable name are the same for the build script and the vscode debug the next run will remove the file.

Can you give an example that can elaborate this?

**Comment by firelizzard18** [comment]
You could set up a build task that will run `go build ./your/exe/package`, which you could then run with `Ctrl+Shift+B` or `Cmd+Shift+B`, followed by `Return`.
